### PR TITLE
[5.4] Fix: FileFactory::generateImage()

### PR DESCRIPTION
**Pull request by sahibalejandro**
This PR fix issue #18289

### CODE REVIEW

**Comment by adamwathan** [comment]
Super weird! Your test passes for me even without the change.

I was actually intentionally storing the temp file on the `File` class only to prevent it from being closed by keeping a reference to it, but you're right it seems like `imagepng` is supposed to close it automatically. That doesn't seem to be happening on my system, but clearly it happens on yours!

If we are going to ditch tmpfile for images, we should probably rework a few other things, like just taking a the `$path` in the file constructor instead of the whole temp file resource, since we don't need a reference to it anymore, and not using `tmpfile` for non-images either, so that everything works in a more consistent way.

File: src/Illuminate/Http/Testing/FileFactory.php [hunk 1]
```diff
@@ -40,8 +40,9 @@ public function image($name, $width = 10, $height = 10)
      */
     protected function generateImage($width, $height)
     {
-        return tap(tmpfile(), function ($temp) use ($width, $height) {
-            imagepng(imagecreatetruecolor($width, $height), $temp);
-        });
+        $path = tempnam(sys_get_temp_dir(), 'kys');
```
Comment (adamwathan) [inline src/Illuminate/Http/Testing/FileFactory.php]: Any reason for this particular prefix?

Comment (sahibalejandro) [inline src/Illuminate/Http/Testing/FileFactory.php]: Nah, just a personal easter egg, nothing important.

**Comment by sahibalejandro** [comment]
So, should I push more commits to get rid of `tmpfile()` and pass a path string in the constructor?

File: src/Illuminate/Http/Testing/FileFactory.php (hunk 1 above)
Comment (ntzm) [inline src/Illuminate/Http/Testing/FileFactory.php]: Probably remove it or change it, 'kys' means 'kill yourself' 😬

Comment (sahibalejandro) [inline src/Illuminate/Http/Testing/FileFactory.php]: omg I didn't realise that 😰
Definitely that is not what it's meant for. Sorry, I'll remove that.

**Comment by adamwathan** [comment]
@sahibalejandro One other approach we might want to consider, you could use an output buffer to capture the output of imagepng and write that to the tempfile, similar to what's mentioned here:

http://php.net/manual/en/function.imagepng.php#28775

If we write the contents to the temp file ourselves, then we could make sure it doesn't get closed.

It might still be better to ditch `tmpfile()` for creating a file manually with `tempnam()` but just an option. I like that with `tmpfile` we don't have to bikeshed over what 3 letter prefix to use 😊

**Comment by adamwathan** [comment]
Changes look great! Maybe use `tmp` for the prefix?

**Comment by sahibalejandro** [comment]
Sure, no problem there.

**Comment by sahibalejandro** [comment]
It's fixed in https://github.com/laravel/framework/commit/79c288f82b058579984ad065f6012b8e26f1ef95

Should I close this PR ?

**Comment by taylorotwell** [comment]
Fixed.

**Pull request closed without merging**